extensions/amp-story/1.0/amp-story.js

```diff
@@ -2237,13 +2240,35 @@ export class AmpStory extends AMP.BaseElement {
 
     const pagesByDistance = this.getPagesByDistance_();
 
-    this.mutateElement(() => {
+    const preloadAllPages = () => {
       pagesByDistance.forEach((pageIds, distance) => {
         pageIds.forEach((pageId) => {
           const page = this.getPageById(pageId);
           page.setDistance(distance);
         });
       });
+    };
+
+    this.mutateElement(() => {
+      if (
+        !isExperimentOn(this.win, 'amp-story-load-first-page-only') ||
+        !prioritizeActivePage
+      ) {
+        return preloadAllPages();
+      }
+      Services.performanceFor(this.win).addEnabledExperiment(
+        'amp-story-load-first-page-only'
+      );
+
+      const activePageId = devAssert(pagesByDistance[0][0]);
+      new Promise((res, rej) => {
+        const page = this.getPageById(activePageId);
+        page.setDistance(0);
+        page.signals().whenSignal(CommonSignals.LOAD_END).then(res);
+        this.storeService_.subscribe(StateProperty.CURRENT_PAGE_ID, rej);
```

Comment:
Can you explain why this might reject? And what behavior should happen if it does reject?

Reply:
Subsequent page navigations will call `preloadAllPages()` too. If the scheduled call to `preloadAllPages` hasn't ran yet, rejecting the promise will ensure it never runs.

Reply:
(Or are you asking for a code comment?)

Reply:
Comments would be good.

Will navigating to the next document cause multiple preload calls? Eg, I stared on page A and preload it, then click to page B, which skips A's preloadAll. But now I'm preloading B, and when that finishes it'll preloadAll and that includes page A.

A map structure recording which pages have already started preloading, could be used, and allow you to cascade outwards from the current page.

Reply:
what "preload" does is it positions the page by the viewport, so the AMP resources call the layoutCallbacks. Calling preload multiple times is totally fine with this implementation

Reply:
Will add a comment 👍

As Gabriel said, what matters is what the last call to "preloadAllPages" is, so we "race" the page loading and page navigation so that on page navigation, any original `preloadAllPages` don't get executed (because the page navigation will execute `preloadAllPages` for the updated pages).